drivers: sam: entropy: implement get_entropy_isr

ioannisg opened the pull request:
The commit contributes the implementation of
get_entropy_isr API function for the SAM entropy
driver. The implementation is similar to get_entropy,
with the difference that it does not invoke k_yield()
when called with the ENTROPY_BUSYWAIT options flag set.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

Fixes #13935

aurel32 reviewed (approved):
Adding `get_entropy_isr()` means that the entropy can be fetched from two different code paths, which means that it adds a race condition. If an ISR happens between the time `entropy_sam_wait_ready()`
returns and `trng->TRNG_ODATA` is read, the same random data will be returned to both code paths.

All that said the problem actually already exists as a thread might be preempted by another one with exactly the same effect. Therefore I believe your changes are perfectly acceptable.

In the long term, we should probably wait for entropy and fetch the random data from a single place, maybe using the RNG interrupt.

ioannisg commented:
@dleach02 could you review this bug-fix PR?

dleach02 commented on drivers/entropy/entropy_sam.c:
```diff
@@ -18,7 +18,7 @@ struct trng_sam_dev_cfg {
 
-static int entropy_sam_wait_ready(Trng *const trng)
+static int entropy_sam_wait_ready(Trng * const trng, u32_t flags)
 {
```
I think the problem with your implementation/changes is that the original intent of entropy_get_entropy_isr() when the blocking flag is zero is to just get as much entropy data is available then return where the return value is the number of bytes retrieved. Note that this may not be equal to the number of bytes requested so it is up to the caller to request the rest.

dleach02 reviewed (commented):
I get that this change is trying to have the underlying support for entropy_get_entropy() and entropy_get_entropy_isr() use a common function but I believe the entropy_get_entropy_isr() function with blocking flag set to 0 is supposed to return what ever data is available without having to wait on the hardware. Now I'm not sure how the sam hardware works in this regard so maybe this is okay. The Nordic implementation appears to want to avoid waiting on the hardware so their code will grab what ever is available and return the count retrieved.

ioannisg commented:
> I get that this change is trying to have the underlying support for entropy_get_entropy() and entropy_get_entropy_isr() use a common function but I believe the entropy_get_entropy_isr() function with blocking flag set to 0 is supposed to return what ever data is available without having to wait on the hardware. Now I'm not sure how the sam hardware works in this regard so maybe this is okay. The Nordic implementation appears to want to avoid waiting on the hardware so their code will grab what ever is available and return the count retrieved.

@dleach02 thanks a lot for your comments.
Right, so this PR extends the driver API for SAM SOCs, so entropy can be safely returned from an ISR without involving kernel APIs. All the PRs does is, effectively, to _not_ call k_yield() when the get-entropy function is invoked by ISR.

Regarding your concern: here is what I see documented in the generic entropy API:
```
/**
 * @brief Fills a buffer with entropy in a non-blocking or busy-wait manner.
 * 	  Callable from ISRs.
 *
 * @param dev Pointer to the device structure.
 * @param buffer Buffer to fill with entropy.
 * @param length Buffer length.
 * @param flags Flags to modify the behavior of the call.
 * @retval number of bytes filled with entropy or -error.
 */
```
The way I have interpreted this is that` entropy_get_entropy_isr()` is _allowed_ to busy-wait inside the ISR. What it is not allowed, is to invoke kernel functions, when the kernel is not initialized, and that's why I 've basically, just removed k_yield().

I understand that the implementation of entropy_get_entropy_isr() may be SOC-specific, and I agree we should try to have some consistence between the SOC families (as much as the hardware differences allow, at least).

Would you like to me to file an issue, to start a discussion to unify the get_entropy_ISR API implementations?

dleach02 commented:
Well this version of the entropy_get_entropy_isr() will still get the whole request unless there is a timeout. I still believe the initial intent of the function was to avoid blocking on the underlying hardware. The changes here do avoid the yield() call so in that sense it meets what was in the description but I still think the implementation is counter to the original intent of entropy_get_entropy_isr() so that should be considered.

ioannisg committed:
drivers: sam: entropy: implement get_entropy_isr

The commit contributes the implementation of
get_entropy_isr API function for the SAM entropy
driver. The implementation is similar to get_entropy,
with the difference that it does not invoke k_yield()
when called with the ENTROPY_BUSYWAIT options flag set.
When the function is invoked without the ENTROPY_BUSYWAIT
flag, it simply returns whatever data is available,
without busy waiting on the RNG herdware.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg reviewed (commented):
@dleach02 I have addressed your feedback in this bug-fix PR

ioannisg committed:
drivers: entropy: nrf: fic minor typos

Fix a couple of minor typos in entropy/entropy_nrf5.c.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

pabigot reviewed (approved):
Reviewing the history of the API I agree the intended behavior of `entropy_get_entropy_isr()` without `ENTROPY_BUSY_WAIT` was to return what's available without waiting.

I also agree there's a pre-existing race condition in the implementation.

I don't have hardware to test this, but by inspection I believe it should work and solve the referenced issue.

ioannisg commented:
> 
> 
> Reviewing the history of the API I agree the intended behavior of `entropy_get_entropy_isr()` without `ENTROPY_BUSY_WAIT` was to return what's available without waiting.
> 
> I also agree there's a pre-existing race condition in the implementation.
> 
> I don't have hardware to test this, but by inspection I believe it should work and solve the referenced issue.

Thanks, @pabigot . I was also convinced by  @dleach02 on this.
I have run the affected test on sam_e70_xplained and it now passes.
